### PR TITLE
Eliminate (`Zip`|`Dup`)(`MapKeys`|`ArrayIndices`).

### PR DESCRIPTION
**Pull request by sellout**
These were of very limited use and basically duplicated the purpose of
the `IdStatus` type used for `ShiftedRead`. Adding an `IdStatus`
parameter to `LeftShift` made it possible to eliminate those MapFunc
nodes.

### CODE REVIEW

**Comment by sellout** [comment]
This is assigned to all three of you, because I need you to review the changes to your connector (and possibly submit a PR against this PR or at least tell me how to fix them). I _think_ Spark probably works. CouchBase might, too. MarkLogic definitely has FIXMEs in there.

You should also review the overall changes, though 😉

**Comment by drostron** [comment]
👍🏾 from me and great to see this for the Couchbase implementation. I'll plan on submitting a PR against this PR, but don't hold this up for it.

Also, needs a revision.md.

**Comment by drostron** [comment]
CB: Handle LeftShift idStatus and add limited type handling for Guard: https://github.com/sellout/quasar/pull/4

**Comment by drostron** [comment]
Needs a merge from master.

File: marklogic/src/main/scala/quasar/physical/marklogic/qscript/ShiftedReadPlanner.scala [hunk 1]
```diff
@@ -36,6 +36,8 @@ private[qscript] final class ShiftedReadPlanner[F[_]: NameGenerator: PrologW]
       val includeId = idStatus match {
         case IncludeId => fn.True
         case ExcludeId => fn.False
+        // FIXME: This ain’t right.
+        case IdOnly    => fn.False
```
Comment (wemrysi) [inline marklogic/src/main/scala/quasar/physical/marklogic/qscript/ShiftedReadPlanner.scala]: What kinds of queries would result in `IdOnly` for a `ShiftedRead`?

**Comment by sellout** [comment]
@drostron Yeah, but I'd like to get reviews/fixes from @wemrysi and @rabbitonweb first.

**Comment by wemrysi** [comment]
@sellout Changes LGTM, though we'll need to merge master in order to fix the MarkLogic impl.

**Comment by sellout** [comment]
Ok, _now_ this is properly updated, and ready for more fixes/comments/whatever.

**Comment by drostron** [comment]
`projectIndexWithObjectFlatten.test` is failing against ML.

**Comment by wemrysi** [comment]
@sellout Not sure if you're notified or not, but https://github.com/sellout/quasar/issues/5 has been submitted for `LeftShift` in MarkLogic.

**Comment by sellout** [comment]
Here is the build for the previous commit, since we don’t get to see it after a `[skip ci]` commit: https://travis-ci.org/quasar-analytics/quasar/builds/177170192

**Pull request merged**